### ISSUE TITLE
FR: Beautified tables

### ISSUE THREAD

**Issue by merlinuwe** [platers/obsidian-linter#827]
Thanks for your great plugin.

I have some ugly tables, like this:

```md
| one | two  | three | four |
| --- | --- | --- | --- |
| 1 |✓| |
|2 | |✓| |
|3| |✓|✓|
| 4 | … | | someting else |
| Total | 100 | 20 | 300000000 |
```

Linter could change them to such a beautiful table:

```md
| one   | two | three | four          |
| ----- | --- | ----- | ------------- |
| 1     | ✓   |       |               |
| 2     |     | ✓     |               |
| 3     |     | ✓     | ✓             |
| 4     | …   |       | someting else |
| Total | 100 | 20    | 300000000     |
```

Ups - in this image you better see what I mean:

![image](https://github.com/platers/obsidian-linter/assets/4138869/25968ad4-f3f1-425e-a674-06459e6b9eb5)

At them moment, I use these online pretifiers:

https://prettier.io/playground/
https://codebeautify.org/markdown-formatter

What do you think?

**Comment by pjkaufman** [comment]
I think this is an interesting idea, but we have previously decided against implementing this on the basis that there were plugins to do this already and there was not a good eay of doing this that would not incurr a lot of bug reports.

**Comment by Mikle-Bond** [comment]
This can be achieved via Advanced Tables community plugin

After installing it, edit `.obsidian/plugins/obsidian-linter/data.json` to include custom command:
```json
  "lintCommands": [
    {
      "id": "table-editor-obsidian:format-all-tables",
      "name": "Format all tables in this file"
    }
  ],
```

Vote on [this issue](https://github.com/tgrosinger/advanced-tables-obsidian/issues/190) if you want to make it possible to add this command via UI.

**Comment by merlinuwe** [comment]
Thank you very much. I tried it with my example and it works as expected. But it also works, when I press <Enter> at the end of the table. I think this is because of the Advanded Tables plugin.  

I don't really understand, what the reason is to enter the code into data.json? 

When I paste 3 of my example tables and Linter the note, I get this:

Before lintering:

```md
| one | two  | three | four |
| --- | --- | --- | --- |
| 1 |✓| |
|2 | |✓| |
|3| |✓|✓|
| 4 | … | | someting else |
| Total | 100 | 20 | 300000000 |


| one | two  | three | four |
| --- | --- | --- | --- |
| 1 |✓| |
|2 | |✓| |
|3| |✓|✓|
| 4 | … | | someting else |
| Total | 100 | 20 | 300000000 |


| one | two  | three | four |
| --- | --- | --- | --- |
| 1 |✓| |
|2 | |✓| |
|3| |✓|✓|
| 4 | … | | someting else |
| Total | 100 | 20 | 300000000 |
```

After lintering:

```md

| one   | two | three | four          |
| ----- | --- | ----- | ------------- |
| 1     | ✓   |       |               |
| 2     |     | ✓     |               |
| 3     |     | ✓     | ✓             |
| 4     | …   |       | someting else |
| Total | 100 | 20    | 300000000     |

| one | two  | three | four |
| ---   | --- | --- | ---           |
| ----- | --- | --- | ------------- |
| 1     | ✓   |     |               |
| 2     |     | ✓   |               |
| 3     |     | ✓   | ✓             |
| 4     | …   |     | someting else |
| Total | 100 | 20  | 300000000     |

| one | two  | three | four |
| ---   | --- | --- | ---           |
| ----- | --- | --- | ------------- |
| 1     | ✓   |     |               |
| 2     |     | ✓   |               |
| 3     |     | ✓   | ✓             |
| 4     | …   |     | someting else |
| Total | 100 | 20  | 300000000     |

```

![image](https://github.com/platers/obsidian-linter/assets/4138869/0ea8c83a-a469-4f6b-8a88-2a1a76cfbcc4)

**Comment by pjkaufman** [comment]
The reason to add those lines to the `data.json` would be to have the Linter run that command when you lint the current file.

As for why it is not formatting the tables correctly, I assume that is a bug in advanced tables if you ran format all tables from a specific table.

**Comment by Mikle-Bond** [comment]
It is actually the first time I see problems in the Advanced tables. I experimented a bit, and here's what I found.

- If Linter changed the position of the table (e.g. removes empty lines before it, so it starts on later line than before linting), then we can see artifacts during AT pass. 
- If tables go one after another without anything in-between, they might become corrupted. Interestingly enough, I've seen the behavior similar to the one in screenshots above, and *also* the opposite behavior, when only the first table gets mangled, and the rest are formatted properly.

I'm not keen on TypeScript, but the only idea I have after digging through code is this: AT gets the location of each table from `app.metadataCache.getFileCache(...).sections`. My assumption is that `CacheMetadata` doesn't get updated before the custom commands run, thus providing the AT with outdated positions to operate on. OTOH, Linter doesn't seem to rely on the cache at all (as far as I can search). Even adding timeout before running custom commands didn't update the `CacheMetadata`.

In my experiments, running AT separately from Linter, even via obsidian console, correctly formatted all tables, even in poorly-formatted markdown file. @pjkaufman , is there a way we could run commands before Linter runs? Is there some hidden reason to limit custom commands to run only after linting?

**Comment by pjkaufman** [comment]
@Mikle-Bond , there are a couple of reasons they run after the file is linted.

The most problematic is I cannot wait for custom commands to finish before running the Linter as they are run asynchronously and, if I recall correctly, they do not properly handling waiting for the command to finish running. As such, I cannot run them before the Linter because that would potentially cause modifications to the current file to get overwritten by the Linter making changes as they could happen between when custom commands start and when after the Linter has already read in the current file's contents.

The other problem might not be a real issue. But it is my understanding that there is a time delay in how long it takes modifications of a file's contents in some cases to get populated to the CodeMirror editor of the current file. Thus if there is that time delay when any custom command runs, then the Linter would operate on bad data.

I will check whether or not custom commands can properly be waited for or if they always run async and cannot be properly awaited. If it is the latter, then I cannot support running custom commands before Linting. If it is possible to run custom commands before the Linter I believe this becomes a duplicate of #596 .

**Comment by Mikle-Bond** [comment]
> * Commands that require an editor/active instance of a file would only get run on open files (currently these can only be added manually to the `data.json` since they don't come back in the list of commands from the obsidian API)

Well, the `editorCallback` commands seem to appear in all command dropdowns (including one in Linter settings). I patched AT, and it works. Only `editorCheckCallback` and `checkCallback` commands will be picky about it.

>   * Would need to do something like the following instead of just `executeCommandById`:

I've tried to set a breakpoint in Obsidian Dev Console to see how commands are run

![Obsidian_4WERcZ8iMD](https://github.com/platers/obsidian-linter/assets/5671588/7997e1d7-fd8a-4019-b31c-7f8dee1975fc)

```js
e.checkCallback ? e.checkCallback(!1) : e.callback ? e.callback() : console.error("Command ".concat(e, " did not provide a callback"))
// and editor*Callback is preferred if present
```

Seems almost like the code snippet you've provided. `executeById` chooses the correct callback, and runs it. If it is a `*Check*` callback, it runs it with `false`. From obsidian docs, it is correct to do:

> The `checkCallback` runs twice. First, to perform a preliminary check to determine whether the command can run. Second, to perform the action.
>Since time may pass between the two runs, you need to perform the check during both calls.
src: https://docs.obsidian.md/Plugins/User+interface/Commands

Also, as all the callbacks are not `async` callbacks, I would assume that they are meant to be executed synchronously.

**Comment by Mikle-Bond** [comment]
Hey, @pjkaufman 

Sorry I died in the middle of the Custom Commands discussion. And thank you for implementing that)

As custom commands have official implementation I see a possible solution for this issue too. I ran some tests with Advanced Tables plugin and some ways to lint files break tables, and some do not.

Minimal setup for the tests: a table with two blank lines between it and previous paragraph, and `consecutive-blank-lines` enabled. E.g. like this:

```md
abcd


| foo   | bar   |
| ----- | ----- |
| hello | world |
```

Results of my observations:

| how lint is executed                                        | do tables break |
| ----------------------------------------------------------- | --------------- |
| run command "Lint current file"                             | yes             |
| save file (with "Lint on save" enabled)                     | yes             |
| change focus to another file/pane/split                     | no              |
| RMB on file header and selecting "Lint file"                | yes             |
| RMB on current file in Navigation and selecting "Lint file" | no              |
| Selecting folder with file and selecting "Lint folder"      | no              |

Which means whenever Lint plugin uses `runLinterEditor` instead of `runLinterFile` it will break tables.

The only functional difference between them seems to be an explicit `this.app.vault.modify` here: https://github.com/platers/obsidian-linter/blob/d8f7f0c2bddc8967a22f4f968e772b30de5dcb39/src/main.ts#L324C5-L324C50

My assumption: it refreshes cache for the file, so the custom command receives up-to-date context on file. Would you mind to look into it? Could it be added to the `runLinterEditor`? Or, maybe, it should be added before each custom command, as there might be other command that rely on up-to-date cache?

**Comment by pjkaufman** [comment]
Hey @Mikle-Bond . I am not sure I fully follow. Right now, I am more focused on fixing up a live preview issue which swaps from using editor replace range and just uses `editor.setValue`. This may fix the issue you are seeing.

However I thought the solution that had been suggested for this was to allow users to decide if a custom command would run before or after the Linter logic. Am I mistaken in this? Thanks for helping me better understand what the next steps are.

**Comment by pjkaufman** [comment]
The reason I mention running the logic before the Linter does is that there is a debounce for when things save when changes are made to the current editor. This meams that the cache is not updated for a second or 2 which causes plugins that rely on cache to have a little bit of an issue running right after the Linter does when running on the editor directly. It may be possible to just save the file directly and bypass this, but I would prefer to not have to do that directly when linting the current file if possible.

**Comment by Mikle-Bond** [comment]
> if a custom command would run before or after the Linter logic

Yes, this particular issue would be solved if custom commands were allowed to run before linter. Though I couldn't imagine any other kind of command that would benefit from this. And with this approach, your plugin will probably run with stale cache, just like AT right now. It might not be an issue right now, but should be documented for the future. 

And it doesn't solve the root of the problem. There might be other plugins that expect cache to be valid at least when their commands start. 

> from using editor replace range and just uses `editor.setValue`

I tried to mock this behavior in sandbox vault. That is, to call `setValue` before custom command. Seems to not update `app.metadataCache` in any way. 
I ran this in Obsidian console:

```js
const dowrap = (x,w)=>{const o = x._old || x.editorCheckCallback; x.editorCheckCallback = w(o); x._old = o; }
dowrap(app.commands.editorCommands['table-editor-obsidian:format-all-tables'], 
	(f)=>{
		return (c,e,v)=>{
			console.log(e.getValue()); 
			e.setValue(e.getValue());
			console.log(app.metadataCache.getFileCache(app.workspace.getActiveFile())); 
			return f(c,e,v);
		}
	}
)
```

..and looked at `sections` property. It was still outdated, while `e.getValue()` returns updated document.

I see that docs mention an event emitted when caches are updated: https://docs.obsidian.md/Reference/TypeScript+API/MetadataCache/on_3 
I guess this is official "lawful" way to make sure cache is ready. 

As for the performance concerns: linting the current file (for me) always happens when I want it to be saved, whether I "save" it, ":w" it, or change focus to another file, I'd want it to be written to the drive. Obsidian will save file anyway, so there shouldn't be much of a slowdown if we add one `app.vault.modify` ourself. 
Or, maybe, not just one. It probably still gets into cache and debounces real filesystem writes.

**Comment by Mikle-Bond** [comment]
Whelp. `app.vault.modify` doesn't update cache. Further tests have shown it does nothing to the cache at all. 

The next significant difference between `runLinterFile` and `runLinterEditor` is here https://github.com/platers/obsidian-linter/blob/d8f7f0c2bddc8967a22f4f968e772b30de5dcb39/src/main.ts#L343

It opens file in new leaf. 
Hypothesis: it triggers full file read, and as a consequence, cache is preheated.
Reality: no. I tried to squeeze in a line like 
```js
const file = app.workspace.getActiveFile();
app.workspace.activeLeaf.openFile(file)
  .then(()=>console.log("patched", app.metadataCache.getFileCache(file).sections))
```
..and caches are still old.

At this point I have only two explanations left.
1. Maybe `activeLeaf.openFile` actually checks if the file is the same as the one already opened. Then there's not much we can do about it.
2. Maybe we get lucky scheduling due to `async-lock`. And I got lucky with this command every time I tried. Looks suspicious, but at this point I'm ready to believe it.

If latter is the case, moving lock from `runLinterFile` into `runCustomCommands` (probably inside the `for` loop, if I understand locks correctly) should solve this issue. But I can't reliably verify that from Obsidian console.

If that doesn't help (most probably it doesn't), the `app.metadataCache.on('changed',...)` is the only reliable thing left.

**Comment by pjkaufman** [comment]
Yeah. It is going to likely have to be an on metadata event based on what I have seen. However that would only apply to single file lints unless I am mistaken.

**Comment by pjkaufman** [comment]
I would need to keep track of the file names that have be linted directly as a single file lint and then check metadata updates here:
```ts
this.plugin.app.metadataCache.on('changed', (file, _, cache) => {
    this.manager?.updateCacheOnExternalFrontmatterUpdate(file.path, cache.frontmatter ?? {});
}),
```
It would be something like this.

**Comment by pjkaufman** [comment]
I think I have a fix for this based on the discussions we had here, but I am not sure if it works entirely, so I would appreciate if someone would test the contents of the following `main.zip` to verify that it does indeed work. It worked for the scenario referenced above where there was an issue with the table not being formatted correctly:
[main.zip](https://github.com/platers/obsidian-linter/files/13591466/main.zip)

@Mikle-Bond @merlinuwe

**Comment by merlinuwe** [comment]
> ... formatted correctly: [main.zip](https://github.com/platers/obsidian-linter/files/13591466/main.zip)

Thank you very much. I just tried it with the table of the first comment of this threat (and 3 tables more) and it works. ;-)

![image](https://github.com/platers/obsidian-linter/assets/4138869/a5ce2f12-02e7-406a-8c98-c210ee1b5968)

**Comment by pjkaufman** [comment]
> > ... formatted correctly: [main.zip](https://github.com/platers/obsidian-linter/files/13591466/main.zip)
> 
> Thank you very much. I just tried it with the table of the first comment of this threat (and 3 tables more) and it works. ;-)
> 
> ![image](https://private-user-images.githubusercontent.com/4138869/288852082-a5ce2f12-02e7-406a-8c98-c210ee1b5968.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE5NzMyMTMsIm5iZiI6MTcwMTk3MjkxMywicGF0aCI6Ii80MTM4ODY5LzI4ODg1MjA4Mi1hNWNlMmYxMi0wMmU3LTQwNmEtOGM5OC1jMjEwZWUxYjU5NjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDdUMTgxNTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MjAyOWQyOWVkMGU4YmQ2ZGU4ZDQzNDU1NTdkN2Y4NTA1YWRjOWQyZDRhZmYxODM0MzQwYTUxYmUzYmExZGI4MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.YuKgSbgb-5Q8GZy_Lur67UslidgrggdF60Pie-SWpBI)
> 
> And if I add the missing `h`, it linters again:
> 
> ![image](https://private-user-images.githubusercontent.com/4138869/288852346-10d8468f-2317-4fd5-a6db-3f28ae8c6415.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDE5NzMyMTMsIm5iZiI6MTcwMTk3MjkxMywicGF0aCI6Ii80MTM4ODY5LzI4ODg1MjM0Ni0xMGQ4NDY4Zi0yMzE3LTRmZDUtYTZkYi0zZjI4YWU4YzY0MTUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQUlXTkpZQVg0Q1NWRUg1M0ElMkYyMDIzMTIwNyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyMzEyMDdUMTgxNTEzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NjkyYjFiZDI0NzlkNzI4MGI3ZWVkOGU4YmEwZWUxZjBhZjgwZDVjOGQ5ZjIzNjg2MDU1ZWRiNDA5YWE1MDU4OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.ZlWlSKJnqK7H3vu24rWhMMJvZy9uR8R1l5iMaZ8uZfc)
> 
> Thank you so much!

Yay! I will see about proceeding with this change. I may modify it slightly to handle all custom commands this way for single and multifile lints.

**Comment by redactedscribe** [comment]
> I would appreciate if someone would test the contents of the following main.zip to verify that it does indeed work.

I hadn't noticed this happening yet, but upon testing the new main.js vs the old, [the examples in the comment above](https://github.com/platers/obsidian-linter/issues/827#issuecomment-1657921965) are indeed fixed. The new main.js also works when Advanced Tables _isn't_ set to beautify (pad cells with spaces) which is what I use.

Question: Will the modified main.js get updated via plugin update once a new Linter version is released? Or should I revert to the old main.js now?

**Comment by Mikle-Bond** [comment]
@pjkaufman Thank you for the test build. I tested it too, and when I tried it, it actually exposed faulty behavior _**once**_, and then started to work properly, as people have reported. I tried to find what might've caused that once-in-a-lifetime glitch.

My only hypothesis is that I *somehow* triggered it right after Obsidian loaded, and it *somehow* used old `main.js` _**once**_, before loading modified one. *Even though it showed a notice that all plugins were loaded and all*...

Anyway, after two days I can't reproduce faulty behavior. I give up and call it a day. I figured I should report things, even if I can't understand them, instead of keeping silence and bashing my head against it. 

Final conclusion: It works. You're a good wizard, @pjkaufman 

@redactedscribe, yes, you can keep modified version. Obsidian doesn't track plugins' integrity (yet) and will update plugin just fine.

**Comment by pjkaufman** [comment]
> > I would appreciate if someone would test the contents of the following main.zip to verify that it does indeed work.
> 
> I hadn't noticed this happening yet, but upon testing the new main.js vs the old, [the examples in the comment above](https://github.com/platers/obsidian-linter/issues/827#issuecomment-1657921965) are indeed fixed. The new main.js also works when Advanced Tables _isn't_ set to beautify (pad cells with spaces) which is what I use.
> 
> Question: Will the modified main.js get updated via plugin update once a new Linter version is released? Or should I revert to the old main.js now?

When a new version of the Linter is released, it will override your current `main.js`. This happens when it sees there is an updated manifest version out there. So you should not need to do anything to get a new release.

**Comment by pjkaufman** [comment]
I have merged the change for this into master. This should affect all custom command usage. Please let us know if this does not work on master or in the next release.
